### PR TITLE
batch conversion tools for large multi-journal installs

### PR DESCRIPTION
**Pull request by kaschioudi**

### CODE REVIEW

**Comment by axfelix** [comment]
Nice! Is there documentation of the CLI tool that this adds so I can test before merging?

**Comment by kaschioudi** [comment]
@axfelix : I haven't written any documentation yet. However from the plugin folder, you can run `php batch.php` or `./batch.php` to view the command-line usage.

**Comment by axfelix** [comment]
OK, cool, I should be able to get something written up based on that so you can focus on other dev stuff :)

**Comment by kaschioudi** [comment]
@axfelix : actually below is the usage output.

```
Usage: 
	 ./batch.php <journal name>					Batch convert a specific journal enabled
	 ./batch.php [--all]						Batch convert all enabled journals
	 ./batch.php [--print]						Prints the list of all enabled journals
	 ./batch.php [--list] <comma-separated list of journals>	Batch convert a comma separated list of journals
```

**Comment by axfelix** [comment]
this isn't handled in this PR, is it? https://github.com/kaschioudi/ojs3-markup/issues/52

Might not hurt to just tackle it at the same time...

**Comment by axfelix** [comment]
So far getting this behaviour on a brand-new test install that has only a single journal and a single submission:

    www-data@lib-garnett:~/html/ojs-plugin-cli/plugins/generic/markup$ ./batch.php --all
    ---------------------------------------------------
    *****> JOURNAL: testing <*****
    ---------------------------------------------------
             => The specified journal path, "admin", does not exist.

**Comment by kaschioudi** [comment]
> this isn't handled in this PR, is it? #52 

true. will do.

@axfelix : I will look into the bug. and this makes me realize that I have one more thing to do. So far I am hardcoding the admin username.

**Comment by kaschioudi** [comment]
@axfelix : Look here https://github.com/kaschioudi/ojs3-markup/blob/45-fix/batch.php#L59-L64
On line 59, modify `admin` to your  install's super admin username. I will provide a fix which updates the error message on line 63 and allows users to specify their admin username on the command line.

@asmecher : Should the user specify the super admin username in the command or I should look for it in the database? is there a best practice in place?

**Comment by asmecher** [comment]
@kaschioudi, yes, I'd suggest letting the user specify it in the command line. That's what we've done for other plugins, e.g. OTOH the XML import/export plugin.

File: batch.php
Comment (asmecher) [inline batch.php:1]: This needs the usual code header (copyrights etc).

File: batch.php
Comment (asmecher) [inline batch.php]: I think the `bootstrap.inc.php` include will set up the environment, including the normal functions like `import` (which should replace `require_once`).

File: batch.php
Comment (asmecher) [inline batch.php]: Our conventions use `@var VariableType someDescription`, i.e. the variable name is implicit.

File: batch.php
Comment (asmecher) [reply batch.php]: I see that you're instantiating the plugin below, so I think it's worth relying on `$plugin->import` for loading plugin-related code.

File: batch.php
Comment (asmecher) [inline batch.php]: `$parameters` either needs to be declared above, or better yet, just use `$argv` instead.

File: batch.php
Comment (asmecher) [inline batch.php]: Text is wrong.

File: batch.php
Comment (asmecher) [inline batch.php]: Is the user group used anywhere more? If so, it might not be safe to assume the first one is appropriate. If not, it should be removed.

File: batch.php
Comment (asmecher) [inline batch.php]: Needs error checking if this isn't correctly specified.

File: batch.php
Comment (asmecher) [inline batch.php]: (Better in general not to serialize to an array unless it's needed -- just loop through while `$contexts->next()` provides an object.)

File: batch.php
Comment (asmecher) [inline batch.php]: Needs error checking.

File: batch.php
Comment (asmecher) [inline batch.php]: Use `lowerCamelCase` conventions for locale keys.

File: batch.php
Comment (asmecher) [inline batch.php]: This tool uses a mix of hard-coded English and locale keys -- better to use locale keys throughout.

File: classes/MarkupBatchConversionHelper.inc.php
Comment (asmecher) [inline classes/MarkupBatchConversionHelper.inc.php]: I think your indentation is wrong here and below.

**Review by asmecher — CHANGES_REQUESTED** [review]
A first round of suggestions...

File: batch.php
Comment (kaschioudi) [reply batch.php:1]: Okay. in the copyright, should I use year 2017 or 2018?

Comment (asmecher) [reply batch.php:1]: 2018. We update the end dates every year.

File: batch.php
Comment (kaschioudi) [reply batch.php]: Okay.

File: batch.php
Comment (kaschioudi) [reply batch.php]: Okay.

File: batch.php
Comment (kaschioudi) [reply batch.php]: Okay. I thought this was defined in the `BatchConversionTool` parent class.

File: batch.php
Comment (kaschioudi) [reply batch.php]: updated.

File: batch.php
Comment (kaschioudi) [reply batch.php]: `MarkupConversionHelper` was using the user group internal. but this commit removed the dependency (https://github.com/kaschioudi/ojs3-markup/commit/405b0fc34a57c1ca331f8407ea4204bf8760e463) I will update functions signatures

Comment (kaschioudi) [reply batch.php]: done.

File: batch.php
Comment (kaschioudi) [reply batch.php]: `processOne` will check the context object (cf. https://github.com/kaschioudi/ojs3-markup/blob/21a0f98d9e3745ae670f2a387744edcfff8428e8/batch.php#L167-L170)

File: batch.php
Comment (kaschioudi) [reply batch.php]: same here. `processOne()` will check `$context` object.

Comment (asmecher) [reply batch.php]: Then the `processOne` self-documentation needs to show it'll accept a `null` `$context`.

File: batch.php
Comment (kaschioudi) [reply batch.php]: Here I'm just reusing an existing locale key. More than a dozen of locale keys in this plugin do not follow `lowerCamelCase`. Should I update all of them?

File: batch.php
Comment (kaschioudi) [reply batch.php]: alright. Okay to put English text in de_DE for now?

File: classes/MarkupBatchConversionHelper.inc.php
Comment (kaschioudi) [reply classes/MarkupBatchConversionHelper.inc.php]: thanks. I still can't figure out why Eclipse does this sometimes when I am refactoring

File: batch.php
Comment (kaschioudi) [reply batch.php]: Okay.

File: batch.php
Comment (asmecher) [reply batch.php]: Yes, that would probably be best.

File: batch.php
Comment (asmecher) [reply batch.php]: No, leave the other languages untranslated (unless you know that language and can do it yourself).

File: batch.php
Comment (kaschioudi) [reply batch.php]: `$context` being an object reference (`Context`) null is a valid value.

Comment (asmecher) [reply batch.php]: I don't think so, e.g. an `assert(is_a($context, 'Context'))` would not pass if `$context` is null.

Typically we use syntax like `Context|null` (or occasionally `Context?`) to indicate nullable/optional parameters.

Comment (kaschioudi) [reply batch.php]: okay.

**Comment by axfelix** [comment]
Thanks for this -- is it good to test again?

**Comment by kaschioudi** [comment]
@axfelix: you could give it a try. I have 2 more features to add.
- ability to pass admin username through command-line argument
- ability to set OTS credentials in config.inc.php. (still not sure how such change will integrate with the UI)
-transform locale keys which dont follow camelCase format.

**Comment by asmecher** [comment]
> still not sure how such change will integrate with the UI

I'd suggest disabling the UI fields entirely when the command-line parameters are set. They should take precedence.

**Comment by kaschioudi** [comment]
> I'd suggest disabling the UI fields entirely when the command-line parameters are set. They should take precedence.

Alright. sounds good.

File: batch.php
Comment (kaschioudi) [reply batch.php]: @asmecher some of those locale keys are not longer used in the plugin. Okay to remove them?

Comment (asmecher) [reply batch.php]: Yes, this is a good thing to do in general (in all languages the translations are present in).

**Comment by kaschioudi** [comment]
> I'd suggest letting the user specify it in the command line. That's what we've done for other plugins, e.g. OTOH the XML import/export plugin.

@asmecher : I have looked around and cannot find the code you are referring to. Could you point me to a specific class / function that read the user from command-line?

**Comment by asmecher** [comment]
```
$ php tools/importExport.php NativeImportExportPlugin
Usage: tools/importExport.php NativeImportExportPlugin [command] ...
Commands:
	import [xmlFileName] [journal_path] [user_name] ...
	export [xmlFileName] [journal_path] articles [articleId1] [articleId2] ...
	export [xmlFileName] [journal_path] article [articleId]
	export [xmlFileName] [journal_path] issues [issueId1] [issueId2] ...
	export [xmlFileName] [journal_path] issue [issueId]

Additional parameters are required for importing data as follows, depending
on the root node of the XML document.

If the root node is <article> or <articles>, additional parameters are required.
The following formats are accepted:

tools/importExport.php NativeImportExportPlugin import [xmlFileName] [journal_path] [user_name]
	issue_id [issueId] section_id [sectionId]

tools/importExport.php NativeImportExportPlugin import [xmlFileName] [journal_path] [user_name]
	issue_id [issueId] section_name [name]

tools/importExport.php NativeImportExportPlugin import [xmlFileName] [journal_path]
	issue_id [issueId] section_abbrev [abbrev]
```

See e.g. https://github.com/pkp/ojs/blob/b2ba0814691ddf3408dfd899f6d0799b1943cfa8/plugins/importexport/native/NativeImportExportPlugin.inc.php#L339.

**Comment by kaschioudi** [comment]
thanks @asmecher : here's the commit https://github.com/kaschioudi/ojs3-markup/pull/53/commits/e924336ca6d98437038f82708cb4374e297fe52a

@axfelix @asmecher this is ready for another round of tests.

File: batch.php
Comment (asmecher) [inline batch.php]: Protected members should be prefixed with a `_`.

File: locale/de_DE/locale.xml
Comment (asmecher) [inline locale/de_DE/locale.xml]: It looks like there's quite a bit of English text in the German locale file; this probably isn't something you introduced, but generally speaking it's a good idea to remove anything like this, so that the translator will notice it's not complete when they next run through the translation process.

File: locale/en_US/locale.xml
Comment (asmecher) [inline locale/en_US/locale.xml]: Are these quotes intentional?

File: templates/settingsForm.tpl
Comment (asmecher) [inline templates/settingsForm.tpl]: I think this will lead to double-escaping. The `{fbvElement}` Smarty function takes care of XSS escaping.

Comment (asmecher) [reply templates/settingsForm.tpl]: (Also elsewhere)

**Comment by asmecher** [comment]
@kaschioudi, just a few comments. Let me know when you've reviewed all the review comments and I'll run through a batch test; then we'll be ready to merge this and forward-port to the master branch of this plugin, I think!

File: batch.php
Comment (kaschioudi) [reply batch.php]: https://github.com/kaschioudi/ojs3-markup/pull/53/commits/0b03be41f17e45218af21f84a206b09988d1fff8

File: locale/de_DE/locale.xml
Comment (kaschioudi) [reply locale/de_DE/locale.xml]: https://github.com/kaschioudi/ojs3-markup/pull/53/commits/e2d0eb5b0fe438febff08fb088981851fea31c34

File: locale/en_US/locale.xml
Comment (kaschioudi) [reply locale/en_US/locale.xml]: https://github.com/kaschioudi/ojs3-markup/pull/53/commits/232d00d6223fc4570b9ded593b05d0429fa88a19

File: templates/settingsForm.tpl
Comment (kaschioudi) [reply templates/settingsForm.tpl]: https://github.com/kaschioudi/ojs3-markup/pull/53/commits/2c75cba62c0c30a921066ae95d5516d0c0ba3020

**Comment by kaschioudi** [comment]
@asmecher : thanks for your feedback. I have reviewed your comments and made the changes.

**Comment by axfelix** [comment]
I'll be on a small vacation through Monday, but will test and merge when I get back.

**Comment by axfelix** [comment]
I think the interpreter shebang (or whatever the equivalent is for PHP) somehow got broken between the last time I tested and now, because now when I try to call this without preceding it with `php` I get:

    garnett@lib-garnett:/var/www/html/ojs-plugin-cli/plugins/generic/markup$ ./batch.php
    ./batch.php: line 1: ?php: No such file or directory
    ./batch.php: line 3: /bin: Is a directory
    ./batch.php: line 4: batch.php: command not found
    ./batch.php: line 5: batch.php: command not found
    ./batch.php: line 6: syntax error near unexpected token `('
    ./batch.php: line 6: ` * Copyright (c) 2003-2018 Simon Fraser University'

Calling it with `php batch.php` works fine, however.

**Comment by axfelix** [comment]
garnett@lib-garnett:/var/www/html/ojs-plugin-cli/plugins/generic/markup$ php ./batch.php
    Usage:
             ./batch.php [user_name] <journal name>                         Batch convert a specific journal enabled
             ./batch.php [user_name] [--all]                                Batch convert all enabled journals
             ./batch.php [--print]                                          Prints the list of all enabled journals
             ./batch.php [user_name] [--list] <comma-separated list of journals>Batch convert a comma separated list of journals

Tab formatting appears to be broken in the last line of the above debug output, we should probably indent every line farther out or shorten the <comma-separated list of journals> text block

**Comment by asmecher** [comment]
IIRC I suggested removing the shebang -- our tools don't use it. Suggest updating the documentation to include the explicit call to `php`.

**Comment by axfelix** [comment]
Tab formatting appears to be broken in the last line of the above debug output, we should probably indent every line farther out or shorten the `<comma-separated list of journals>` text block

**Comment by axfelix** [comment]
Oh. Is there anything gained from removing it... or is it just a convention?

**Comment by asmecher** [comment]
Unless Windows supports shebangs, I figured it was better to go without, and have the documentation apply equally for both Windows and *NIX OSs.

**Comment by axfelix** [comment]
Oh. I think most people doing actual CLI work on Windows these days will be using a bash shell anyway, so I don't think that necessarily applies...

**Comment by axfelix** [comment]
OK, we probably also need slightly better error reporting (I can add it myself after merging as soon as I know what's actually going on here):

    garnett@lib-garnett:/var/www/html/ojs-plugin-cli/plugins/generic/markup$ php ./batch.php otsalpha
    PHP Notice:  Undefined offset: 0 in /var/www/html/ojs-plugin-cli/plugins/generic/markup/batch.php on line 129
    The specified journal path does not exist.
    garnett@lib-garnett:/var/www/html/ojs-plugin-cli/plugins/generic/markup$ php ./batch.php otsalpha -all
    The specified journal path does not exist.
    garnett@lib-garnett:/var/www/html/ojs-plugin-cli/plugins/generic/markup$ php ./batch.php otsalpha --all
    ---------------------------------------------------
    *****> JOURNAL: testing <*****
    ---------------------------------------------------
    => Processing submission: Test Submission
    
    1 submission(s) found & 0 processed!

The first couple erroneous runs due to me mis-typing the syntax should probably have an error message other than "The specified journal path does not exist." -- perhaps instead "Please verify the input syntax -- it looks like you specified a journal path that doesn't exist." And I'm not actually clear on why, in the successful run, I got 1 submission found & 0 processed.

**Comment by asmecher** [comment]
> Oh. I think most people doing actual CLI work on Windows these days will be using a bash shell anyway, so I don't think that necessarily applies...

I have no idea what people are using in Windows these days, I just didn't see a downside to making the `php` call explicit for Windows shell users (= the majority of beginner XAMPP users, I'd expect). If it's going to bug you, feel free to re-add it -- my opinion isn't strong on this.

**Comment by axfelix** [comment]
OK, yeah, I'm gonna re-add it at some point. In fairness I am sort of fanatically cross-platform (I'm using Windows right now) but my opinion on Windows support for this stuff in 2018 is that you shouldn't go one iota out of your way for it, especially if it means (marginally) degrading the experience for *nix users, because all of the good stacks (mingw/Git bash, nginx, etc.) have a Windows distribution these days anyway and everyone should just be using those -- which do implicitly support all the *nix syntax, as that's their entire purpose -- rather than encouraging people to learn Microsoft stacks in lousier environments and making their life tougher from the get-go.

**Comment by kaschioudi** [comment]
> "Please verify the input syntax -- it looks like you specified a journal path that doesn't exist." 

I am currently running a multi journal test batch conversion . I will make this change tomorrow. 

> And I'm not actually clear on why, in the successful run, I got 1 submission found & 0 processed.

Basically, this means there is one submission in the journal and none was processed probably because  the script was unable to automatically guess which file to convert. Please feel free to modify the wording if you find it confusing.

**Comment by axfelix** [comment]
Got it. What's the current logic for automatically guessing a file to convert? Again, I'm happy to write the documentation if I know what I'm starting from.

**Comment by axfelix** [comment]
OK, added documentation and just about ready to merge this after a tiny bit more testing if everyone else feels good with where we're at! Thanks all.

**Comment by axfelix** [comment]
I think it should probably be printing "Pending" to the terminal a little less (even if it checks every few seconds, it quickly fills up a terminal), maybe not more than once every 30s?

Regardless, this is working great, I'm merging now!

**Pull request merged**
